### DISCUSSION TITLE
How to call `refresh` of `useFetch` immediately in `onMounted`?

### DISCUSSION THREAD

**Discussion by Dylan0916** [nuxt/nuxt#25934] [Questions] [answered]
Hi, I have a page that needs to take care of SEO, so I use `useFetch` to fetch data.
the API returns different data for some fields when there is auth info. due to architectural constraints, I cannot get auth info on the server side, so I need to make an additional request on the client side.

However, I found that calling `refresh()` in `onMounted` does not have any effect unless I put it in a `setTimeout`. why is this happening? Is there a more elegant solution?

Thank you.

Below is my sample code

```javascript
const { data, refresh } = await useFetch('url');

onMounted(() => {
  refresh();
});
```

**Comment by TheAlexLichter** [comment]
Why not using `server: false`? `const { data, refresh } = await useFetch('url', { server: false });`

**Reply by Dylan0916** [reply]
Thanks for the reply. because I need to consider SEO issues, I need data on the server side and cannot use `server: false` 😓

**Comment by EddyDezuraud** [comment]
I believe the second request triggered by `onMounted` fires too early, even before the response from the first `useFetch` is rendered. I think you could try adding `immediate: false` as an option in the `useFetch` to prevent the request from firing before `onMounted`.

👉 [useFetch documentation](https://nuxt.com/docs/api/composables/use-fetch#params)


```js
const { data, refresh } = await useFetch('url', {immediate: false});

onMounted(() => {
  refresh();
});
```

**Reply by Dylan0916** [reply]
Thanks for the reply. as far as I understand, after using `await`, the page will wait for the api to return before starting to execute the client side code (`onMounted`). not sure if I understand it correctly

but I need to consider SEO issues, I need data on the server side and cannot use `immediate: false`

**Reply by TheAlexLichter** [reply]
If you need SEO, why would you refresh the data on mounted? 🤔

**Reply by Dylan0916** [reply]
Because there is user info when the request on the client side, but not on the server side, that's why I need to make another request on the client side.

**Reply by EddyDezuraud** [reply]
If I understand correctly, you want to make an initial server-side request (SSR) to preload the data for SEO purposes. Then, you'd like to make a second client-side request, presumably passing client information as parameters (to modify elements on the interface)?

**Reply by Dylan0916** [reply]
Yes, that's right, this is my purpose

**Reply by EddyDezuraud** [reply, marked as answer]
I believe we could provide a more precise and optimized answer with a concrete example from your code. However, I see several possible approaches to solve your problem.

### 1st possibility:
Utilize a computed property called readyForClientFetch, to which you map a watcher that refreshes your call when onMounted has been invoked and the first useFetch is completed.

Example:

```js
const { data, refresh, status } = await useFetch('url');

const isMounted = ref(false);

const readyForClientFetch = computed(() => status.value === 'success' && isMounted.value)

onMounted(() => {
  isMounted.value = true;
});

watch(readyForClientFetch.value,
  (newVal, oldVal) => {
        if(newVal === true) {
          refresh();
        }
  },{ immediate: true });
```

### 2nd possibility:
The second possibility could be to use the watch property of useFetch to automatically trigger a new call.

Example:
```js
const isMounted = ref(false);

const { data, refresh, status } = await useFetch('url', {watch: [isMounted]});

onMounted(() => {
  isMounted.value = true;
});
```


I'm not an expert in using UseFetch in SSR mode, so there might be even more optimized solutions out there. However, I believe these approaches should help you address your issue.

**Reply by Dylan0916** [reply]
Thank you very much, the 1st works!
